karoo_gp/base_class.py

```diff
@@ -675,32 +447,162 @@ def fx_data_params_write_json(self):
+                    fittest_dict[tree.id] = tree.expression
+                    last_fittest = tree
+        else:
+            for tree in trees:
+                if tree.fitness == self.perfect_score:
+                    fittest_dict[tree.id] = tree.expression
+        return fittest_dict
+
+class MultiClassifier_GP(Base_GP):
+
+    def __init__(self, *args, **kwargs):
+        """Add kernel-specific scoring function(s) and kwargs"""
+
+        def n_correct(y_true, y_pred):
+            """Default classification fitness: number of correct predictions"""
+            return float(np.sum(y_true == y_pred))
+
+        def cls_report_zero_div(y_true, y_pred):
+            # TODO: cli test expected output (from '..write_json()')
+            # classification report dict keys are floats. Should use int.
+            return skm.classification_report(
+                y_true.astype(np.float32),
+                y_pred.astype(np.float32),
+                zero_division=0, output_dict=True)
+
+        def conf_matrix_as_list(y_true, y_pred):
+            return skm.confusion_matrix(y_true, y_pred).tolist()
+
+        kwargs['scoring'] = dict(fitness=n_correct,
+                                 classification_report=cls_report_zero_div,
+                                 confusion_matrix=conf_matrix_as_list)
+        kwargs['higher_is_better'] = True
+        super().__init__(*args, **kwargs)
+
+    def fit(self, X, y, *args, **kwargs):
+        """Initialize encoder to be used by scorer"""
+        self.encoder = self.encoder or LabelEncoder(y)
+        super().fit(X, y, *args, **kwargs)
+
+class LabelEncoder:
+    """
+    For classification tasks, classes are encoded in sample data as 0-N
+    integers, where N is the number of classes (e.g. [0, 1, 2, 3], N = 4).
+    When calculating tree output, y values are skewed so that they're centered
+    at 0 (e.g. [-1.5, -0.5, 0.5, 1.5], skew = -1.5).
+
+    This class stores the skew value and implements encode/decode methods:
+      encode (y -> output): adds skew to input
+      decode (output -> y): subtract skew, round, clip to 0-N
+    """
+    def __init__(self, y=None, n_classes=None):
+        self.n_classes = n_classes or len(np.unique(y))
+        self.skew = (self.n_classes / 2) - 0.5
+
+    def encode(self, x):
+        """Convert 0-N encoding to zero-centered integer encoding"""
+        return x - self.skew
+
+    def decode(self, x):
+        """Convert zero-centered back to 0-N and crop tails"""
+        if len(x.shape) > 1:  # Use recursion for batch processing
+            output = np.zeros(x.shape, dtype=np.int32)
+            for i in range(x.shape[0]):
+                output[i] = self.decode(x[i])
+            return output
+        else:   # Decode 1-dim array
+            output = x + self.skew
+            output = np.where(output % 1 <= 0.5,    # Round to nearest int
+                np.floor(output), np.ceil(output))  # 0.5 always rounds down
```

Comment:
Isn't there a better way of doing this?

Reply:
numpy has a few rounding functions, but from what I've seen they don't allow you to specify the rounding rules (like `decimal` does).
I found https://stackoverflow.com/a/60053837 that suggests this:
```py
>>> a = np.array([1, 1.5, 2, 2.5, 3, 3.5])
>>> np.rint(np.nextafter(a, a-1))
array([1., 1., 2., 2., 3., 3.])
```
Is there a reason to deviate from the standard rounding, and also for wanting `0.5` to round down instead of up?

Reply:
The rounding affects the output, so if we change the rounding scheme it will change the tests. 

The original was 0.5-down. In the papers on Binary trees they usually say <= 0 is False, >0 is true, which is conceptually the same approach.